keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/ForwardSampler.java

```diff
@@ -0,0 +1,114 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ForwardSampler {
+
+    private static final double LOG_PROB_OF_PRIOR = 0.;
+    private static final ProposalDistribution PRIOR_PROPOSAL = new PriorProposalDistribution();
+
+    public ForwardSampler() {
+    }
+
+    /**
+     * Samples from a Bayesian Network.
+     * Samples are taken by sampling from the prior of the desired variables.
+     * Probabilistic variables are sampled first.
+     * Non probabilistic variables are computed second.
+     *
+     * @param network the network to sample from
+     * @param fromVariables the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    public NetworkSamples sample(BayesianNetwork network,
+                                        List<? extends Variable> fromVariables,
+                                        int sampleCount) {
+        return sample(network, fromVariables, sampleCount, KeanuRandom.getDefaultRandom());
+    }
+
+    public NetworkSamples sample(BayesianNetwork network,
+                                        List<? extends Variable> fromVariables,
+                                        int sampleCount,
+                                        KeanuRandom random) {
+
+        Preconditions.checkArgument(network.getObservedVertices().size() == 0, "Cannot forward sample from a network with observations");
+
+        List<Vertex> fromVertices = fromVariables.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+
+        Set<Vertex> probabilisticSubset = fromVertices.stream().filter(v -> v instanceof Probabilistic).collect(Collectors.toSet());
+        Set<Vertex> nonProbabilisticSubset = fromVertices.stream().filter( v -> !(v instanceof Probabilistic)).collect(Collectors.toSet());
+
+        upstreamOfProbabilisticDoesNotContainProbabilistic(probabilisticSubset);
+
+        ComputableModel graph = new KeanuComputableModel(new HashSet<>(network.getAllVertices()));
+        KeanuProposalApplicationStrategy proposalApplicationStrategy = new KeanuProposalApplicationStrategy(probabilisticSubset);
+
+        return sample(graph, probabilisticSubset, nonProbabilisticSubset, proposalApplicationStrategy, sampleCount, random);
+    }
+
+    private NetworkSamples sample(ComputableModel graph,
+                                        Set<? extends Variable> probabilisticFromVariables,
+                                        Set<? extends Variable> nonProbabilisticFromVariables,
+                                        ProposalApplicationStrategy proposalApplicationStrategy,
+                                        int sampleCount,
+                                        KeanuRandom random) {
+
+        Set<? extends Variable> allSampleFromVariables = Sets.union(probabilisticFromVariables, nonProbabilisticFromVariables);
```

Comment:
Do we need to generate this again?  Couldn't we just pass it through from the function above?